README.md
# Getting Started

This project was created to offer an easier way to use the templating service from Morphotech
 in other Java projects. 


# Plato Service Helper (Java)

## Maven

```xml
<properties>
    <plato-service-helper.version>2.0.1</plato-service-helper.version>
</properties>
...
<dependency>
    <groupId>com.morphotech</groupId>
    <artifactId>plato-helper-java</artifactId>
    <version>${plato-service-helper.version}</version>
</dependency>
```

## Configuration

#### Recommended creating a service (or similar)

```java
@Configuration
public class PlatoConfig {

    @Bean
    public PlatoService platoService(
        @Value("${plato-service.base-url}") String baseUrl) {
        return new PlatoService(baseUrl);
    }
}
```


CHANGELOG.md
# Changelog

## Version 1.0.0 - 2022-10-21

### Task

First version released.

src/test/java/com/morphotech/BuilderUtils.java
package com.morphotech;

public class BuilderUtils {

    public static PlatoService CreateBasicTemplatingService(
            String baseUrl,
            String tokenUrl,
            String clientId,
            String secret
    ){
        return new PlatoService(baseUrl, tokenUrl, clientId, secret);
    }

}


src/test/java/com/morphotech/PlatoHelperJavaApplicationTest.java
package com.morphotech;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Test;

import javax.net.ssl.SSLSession;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpHeaders;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.Map;
import java.util.Optional;
import java.util.Random;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.notNullValue;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class PlatoHelperJavaApplicationTest {

    private final HttpClient httpClient =
            mock(HttpClient.class);

    @Test
    void getTemplateExample_TEXT_HTML() throws Exception {

        byte[] b = byteArrayBuilder(20);
        var httpResponseByteArray = httpResponseBuilder(200, b);

        when(httpClient.send(any(), any()))
                .thenReturn(httpResponseByteArray);

        var templatingService = new PlatoService(httpClient, "http://localhost.com");

        var response = templatingService.getTemplateExample("template-id", MediaType.TEXT_HTML);

        assertThat(response, notNullValue());
        assertThat(response.length, is(20));
    }

    @Test
    void getAllTemplates() throws Exception {
        var json = jsonBuilder(Map.of(
                "template_schema", "value",
                "field1", "value1",
                "field2", "value2"));

        var httpResponseString = httpResponseBuilder(200, json);

        when(httpClient.send(any(), any()))
                .thenReturn(httpResponseString);

        var templatingService = new PlatoService(httpClient, "http://localhost.com");

        var response = templatingService.getAllTemplates();
        assertThat(response.contains("template_schema"), is(true));

    }

    @Test
    void composeTemplatingFile() throws Exception {
        byte[] b = byteArrayBuilder(30);
        var httpResponseByteArray = httpResponseBuilder(200, b);

        when(httpClient.send(any(), any()))
                .thenReturn(httpResponseByteArray);

        var templatingService = new PlatoService(httpClient, "http://localhost.com");


        var jsonToPost = jsonBuilder(Map.of(
                "recipient_name", "Subject 37",
                "certificate_number", "C12345"));

        var response = templatingService.composeTemplateById(
                "template-id", MediaType.APPLICATION_PDF, jsonToPost
        );

        assertThat(response, notNullValue());
        assertThat(response.length, is(30));
    }

    // helpers

    private byte[] byteArrayBuilder(int size) {
        var b = new byte[size];
        new Random().nextBytes(b);
        return b;
    }

    private String jsonBuilder(Map<String, String> map) throws JsonProcessingException {
        ObjectMapper mapper = new ObjectMapper();
        return mapper.writeValueAsString(map);
    }

    private HttpResponse httpResponseBuilder(int status, Object body) {

        return new HttpResponse() {
            @Override
            public int statusCode() {
                return status;
            }

            @Override
            public HttpRequest request() {
                return null;
            }

            @Override
            public Optional<HttpResponse> previousResponse() {
                return Optional.empty();
            }

            @Override
            public HttpHeaders headers() {
                return null;
            }

            @Override
            public Object body() {
                return body;
            }

            @Override
            public Optional<SSLSession> sslSession() {
                return Optional.empty();
            }

            @Override
            public URI uri() {
                return null;
            }

            @Override
            public HttpClient.Version version() {
                return null;
            }
        };
    }
}


src/test/java/com/morphotech/TemplateHelperJavaApplicationTest.java
package com.morphotech;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.morphotech.exception.TemplatingServiceException;
import org.junit.jupiter.api.Test;

import javax.net.ssl.SSLSession;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpHeaders;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.Map;
import java.util.Optional;
import java.util.Random;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.notNullValue;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class TemplateHelperJavaApplicationTest {

    private HttpClient httpClient =
            mock(HttpClient.class);

    @Test
    void getTemplateExample_TEXT_HTML() throws Exception {

        var json = jsonBuilder(Map.of(
                "access_token", "abcvdef",
                "expires_in", "300"));
        var httpResponseToken = httpResponseBuilder(200, json);

        byte[] b = byteArrayBuilder(20);
        var httpResponseByteArray = httpResponseBuilder(200, b);

        when(httpClient.send(any(), any()))
                .thenReturn(httpResponseToken, httpResponseByteArray);

        var templatingService = new TemplatingService(httpClient,
                "http://localhost.com",
                "http://localhost.com/auth/realms/micro-keycloak/protocol/openid-connect/token",
                "template-client-id",
                "this-is-secret");

        var response = templatingService.getTemplateExample("template-id", MediaType.TEXT_HTML);

        assertThat(response, notNullValue());
        assertThat(response.length, is(20));
    }

    @Test
    void getAllTemplates() throws Exception {
        var json = jsonBuilder(Map.of(
                "access_token", "abcvdef",
                "expires_in", "300"));
        var httpResponseToken = httpResponseBuilder(200, json);

        json = jsonBuilder(Map.of(
                "template_schema", "value",
                "field1", "value1",
                "field2", "value2"));

        var httpResponseString = httpResponseBuilder(200, json);

        when(httpClient.send(any(), any()))
                .thenReturn(httpResponseToken, httpResponseString);

        var templatingService = new TemplatingService(httpClient,
                "http://localhost.com",
                "http://localhost.com/auth/realms/micro-keycloak/protocol/openid-connect/token",
                "template-client-id",
                "this-is-secret");

        var response = templatingService.getAllTemplates();
        assertThat(response.contains("template_schema"), is(true));

    }

    @Test
    void composeTemplatingFile() throws Exception {
        var json = jsonBuilder(Map.of(
                "access_token", "abcvdef",
                "expires_in", "300"));
        var httpResponseToken = httpResponseBuilder(200, json);

        byte[] b = byteArrayBuilder(30);
        var httpResponseByteArray = httpResponseBuilder(200, b);

        when(httpClient.send(any(), any()))
                .thenReturn(httpResponseToken, httpResponseByteArray);

        var templatingService = new TemplatingService(httpClient,
                "http://localhost.com",
                "http://localhost.com/auth/realms/micro-keycloak/protocol/openid-connect/token",
                "template-client-id",
                "this-is-secret");


        var jsonToPost = jsonBuilder(Map.of(
                "recipient_name", "Subject 37",
                "certificate_number", "C12345"));

        var response = templatingService.composeTemplateById(
                "template-id", MediaType.APPLICATION_PDF, jsonToPost
        );

        assertThat(response, notNullValue());
        assertThat(response.length, is(30));
    }

    @Test
    void getAccessTokenNull_throw_exception() throws IOException, InterruptedException {

        when(httpClient.send(any(), any()))
                .thenReturn(null);

        var exceptionThrown = assertThrows(TemplatingServiceException.class, () -> {
            new TemplatingService(httpClient,
                    "http://localhost.com",
                    "http://localhost.com/auth/realms/micro-keycloak/protocol/openid-connect/token",
                    "template-client-id",
                    "this-is-secret");
        });

        assertThat(exceptionThrown.getMessage(), is("Failed to get access token for template service"));
    }

    @Test
    void getTemplateExample_TEXT_HTML_getNewAccessToken() throws Exception {
        // Given
        var json = jsonBuilder(Map.of(
                "access_token", "abcvdef",
                "expires_in", "300"));
        var httpResponseToken = httpResponseBuilder(200, json);

        var httpResponse_401 = httpResponseBuilder(401, null);

        var json_new_token = jsonBuilder(Map.of(
                "access_token", "poiuygdsd",
                "expires_in", "300"));
        var httpResponseToken_new_token = httpResponseBuilder(200, json_new_token);

        byte[] b = byteArrayBuilder(20);
        var httpResponseByteArray = httpResponseBuilder(200, b);



        // When
        when(httpClient.send(any(), any()))
                .thenReturn(httpResponseToken, httpResponse_401, httpResponseToken_new_token, httpResponseByteArray);

        // Do request
        var templatingService = new TemplatingService(httpClient,
                "http://localhost.com",
                "http://localhost.com/auth/realms/micro-keycloak/protocol/openid-connect/token",
                "template-client-id",
                "this-is-secret");

        var response = templatingService.getTemplateExample("template-id", MediaType.TEXT_HTML);

        assertThat(response, notNullValue());
        assertThat(response.length, is(20));
    }

    // helpers

    private byte[] byteArrayBuilder(int size) {
        var b = new byte[size];
        new Random().nextBytes(b);
        return b;
    }

    private String jsonBuilder(Map<String, String> map) throws JsonProcessingException {
        ObjectMapper mapper = new ObjectMapper();
        return mapper.writeValueAsString(map);
    }

    private HttpResponse httpResponseBuilder(int status, Object body) {

        return new HttpResponse() {
            @Override
            public int statusCode() {
                return status;
            }

            @Override
            public HttpRequest request() {
                return null;
            }

            @Override
            public Optional<HttpResponse> previousResponse() {
                return Optional.empty();
            }

            @Override
            public HttpHeaders headers() {
                return null;
            }

            @Override
            public Object body() {
                return body;
            }

            @Override
            public Optional<SSLSession> sslSession() {
                return Optional.empty();
            }

            @Override
            public URI uri() {
                return null;
            }

            @Override
            public HttpClient.Version version() {
                return null;
            }
        };
    }

}


src/main/java/com/morphotech/PlatoService.java
package com.morphotech;

import com.morphotech.exception.TemplatingServiceException;
import com.morphotech.exception.WebServiceException;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.Map;

import static java.util.Collections.EMPTY_MAP;

public class PlatoService {

    private static final String ACCEPT_HEADER = "Accept";
    private static final String CONTENT_TYPE_HEADER = "Content-Type";
    private static final String CONTENT_APPLICATION_JSON = "application/json";
    private static final Integer DEFAULT_REQUEST_TIMEOUT = 15;
    private final String baseUrl;
    private HttpClient httpClient;
    private int requestTimeOut = DEFAULT_REQUEST_TIMEOUT;

    /**
     * For Testing purposes
     */
    public PlatoService(HttpClient httpClient, String baseUrl) {
        this.baseUrl = baseUrl;
        this.httpClient = httpClient;
    }

    public PlatoService(String baseUrl) {
        this.baseUrl = baseUrl;
    }

    public PlatoService(String baseUrl, int requestTimeout) {
        this.baseUrl = baseUrl;
        this.requestTimeOut = requestTimeout;
    }

    private void setupHttpClient() {
        if (httpClient == null) {
            httpClient = HttpClient.newBuilder()
                    .connectTimeout(Duration.ofSeconds(requestTimeOut))
                    .build();
        }
    }

    /**
     * Gets example file based on the template
     *
     * @param templateId [example: vizidox-student]
     * @param mediaType  a type of {@link MediaType}
     * @return example of file based on template
     */
    public byte[] getTemplateExample(String templateId, MediaType mediaType) throws WebServiceException {
        setupHttpClient();

        var request = buildHttpRequestHeader(
                baseUrl + "/template/" + templateId + "/example",
                Map.of(ACCEPT_HEADER, mediaType.getString()),
                RequestMethod.GET
        );

        return makeRequest(request, HttpResponse.BodyHandlers.ofByteArray()).body();
    }

    /**
     * @return template information
     */
    public String getAllTemplates() throws WebServiceException {
        setupHttpClient();

        var request = buildHttpRequestHeader(
                baseUrl + "/templates/",
                EMPTY_MAP,
                RequestMethod.GET
        );

        return makeRequest(request, HttpResponse.BodyHandlers.ofString()).body();
    }

    /**
     * Composes file based on the template
     *
     * @param templateId id of the template on service
     * @param mediaType  determines what kind of file is outputed. One of {@link MediaType}
     * @param schema     body to compose file with, must be according to the template schema
     * @return composed file
     */
    public byte[] composeTemplateById(String templateId, MediaType mediaType, String schema) throws WebServiceException {
        setupHttpClient();

        var request = buildHttpRequestHeader(
                baseUrl + "/template/" + templateId + "/compose",
                Map.of(ACCEPT_HEADER, mediaType.getString(),
                        CONTENT_TYPE_HEADER, CONTENT_APPLICATION_JSON),
                RequestMethod.POST,
                schema
        );

        return makeRequest(request, HttpResponse.BodyHandlers.ofByteArray()).body();
    }

    /**
     * HTTP send request method
     * <p>
     * This method exists so that the application can resend the request with a new access token in case of 401
     *
     * @param request
     * @param bodyHandler
     * @param <T>
     * @return
     */
    private <T> HttpResponse<T> makeRequest(HttpRequest request, HttpResponse.BodyHandler<T> bodyHandler) throws WebServiceException {
        try {

            var response = httpClient.send(request, bodyHandler);

            if (response.statusCode() != 200) {
                throw new WebServiceException("Failed to access Plato Service with http status: " + response.statusCode());
            }

            return response;

        } catch (IOException e) {
            throw new TemplatingServiceException("Failed to send or receive request through HTTP client");
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new TemplatingServiceException("The thread was interrupted. Failed to access Plato Service");
        }
    }

    /**
     * Override of {@link #buildHttpRequestHeader(String, Map, RequestMethod, String)}
     *
     * @param url
     * @param headerMap
     * @param requestMethod
     * @return
     */
    private HttpRequest buildHttpRequestHeader(String url, Map<String, String> headerMap, RequestMethod requestMethod) {
        return buildHttpRequestHeader(url, headerMap, requestMethod, null);
    }

    /**
     * This method exists to make it easier to build the headers of the HTTP request
     * It also validates the requestMethod
     *
     * @param url
     * @param headerMap
     * @param requestMethod
     * @param bodyContent
     * @return
     */
    private HttpRequest buildHttpRequestHeader(String url, Map<String, String> headerMap, RequestMethod
            requestMethod, String bodyContent) {
        var httpRequest = HttpRequest.newBuilder(URI.create(url));

        if (null != headerMap && !headerMap.isEmpty()) {
            headerMap.keySet().forEach(key -> httpRequest.setHeader(key, headerMap.get(key)));
        }

        httpRequest.timeout(Duration.ofSeconds(requestTimeOut));

        switch (requestMethod) {
            case GET:
                return httpRequest.GET().build();
            case POST:
                return httpRequest.POST(HttpRequest.BodyPublishers.ofString(bodyContent)).build();
            default:
                throw new TemplatingServiceException("HTTP Request method not recognized: " + requestMethod);
        }
    }
}


src/main/java/com/morphotech/TemplatingService.java
package com.morphotech;


import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.morphotech.exception.TemplatingServiceException;
import com.morphotech.exception.WebServiceException;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.Map;

public class TemplatingService {

    private final String templatingBaseUrl;

    private final String templatingTokenUrl;

    private final String templatingClientId;

    private final String templatingSecret;

    private HttpClient httpClient;
    private String tokenBearer;

    private static final String ACCEPT_HEADER = "Accept";
    private static final String AUTHORIZATION_HEADER = "Authorization";
    private static final String CONTENT_TYPE_HEADER = "Content-Type";
    private static final String BEARER_HEADER = "Bearer ";

    private final ObjectMapper objectMapper = new ObjectMapper();

    /**
     * For Testing purposes
     */
    public TemplatingService(HttpClient httpClient, String templatingBaseUrl, String templatingTokenUrl, String templatingClientId, String templatingSecret) {
        this.templatingBaseUrl = templatingBaseUrl;
        this.templatingTokenUrl = templatingTokenUrl;
        this.templatingClientId = templatingClientId;
        this.templatingSecret = templatingSecret;

        this.httpClient = httpClient;
        getAccessToken();
    }

    public TemplatingService(String templatingBaseUrl, String templatingTokenUrl, String templatingClientId, String templatingSecret) {
        this.templatingBaseUrl = templatingBaseUrl;
        this.templatingTokenUrl = templatingTokenUrl;
        this.templatingClientId = templatingClientId;
        this.templatingSecret = templatingSecret;
    }

    private void setupHttpClient() {
        if (httpClient == null) {
            httpClient = HttpClient.newBuilder().build();
            getAccessToken();
        }
    }

    // API

    /**
     * Gets example file based on the template
     *
     * @param templateId [example: vizidox-student]
     * @param mediaType  a type of {@link MediaType}
     * @return example of file based on template
     */
    public byte[] getTemplateExample(String templateId, MediaType mediaType) throws WebServiceException {

        var request = buildHttpRequestHeader(
                templatingBaseUrl + "/template/" + templateId + "/example",
                Map.of(ACCEPT_HEADER, mediaType.getString(),
                        AUTHORIZATION_HEADER, BEARER_HEADER + tokenBearer),
                RequestMethod.GET
        );

        return makeRequest(request, HttpResponse.BodyHandlers.ofByteArray()).body();
    }

    /**
     * @return template information
     */
    public String getAllTemplates() throws WebServiceException {

        var request = buildHttpRequestHeader(
                templatingBaseUrl + "/templates/",
                Map.of(AUTHORIZATION_HEADER, BEARER_HEADER + tokenBearer),
                RequestMethod.GET
        );

        return makeRequest(request, HttpResponse.BodyHandlers.ofString()).body();
    }

    /**
     * Composes file based on the template
     *
     * @param templateId id of the template on service
     * @param mediaType  determines what kind of file is outputed. One of {@link MediaType}
     * @param schema     body to compose file with, must be according to the template schema
     * @return composed file
     */
    public byte[] composeTemplateById(String templateId, MediaType mediaType, String schema) throws WebServiceException {

        var request = buildHttpRequestHeader(
                templatingBaseUrl + "/template/" + templateId + "/compose",
                Map.of(ACCEPT_HEADER, mediaType.getString(),
                        AUTHORIZATION_HEADER, BEARER_HEADER + tokenBearer,
                        CONTENT_TYPE_HEADER, "application/json"),
                RequestMethod.POST,
                schema
        );

        return makeRequest(request, HttpResponse.BodyHandlers.ofByteArray()).body();
    }


    // Authentication

    /**
     * Makes request to get Access Token and sets it on local variable
     */
    private void getAccessToken() {

        String tokenRequest =
                "client_id=" + templatingClientId
                        + "&client_secret=" + templatingSecret
                        + "&grant_type=" + "client_credentials"
                        + "&scope=" + "content-provider-scope";

        HttpRequest oAuth2Request = buildHttpRequestHeader(
                templatingTokenUrl,
                Map.of(
                        ACCEPT_HEADER, "application/json",
                        CONTENT_TYPE_HEADER, "application/x-www-form-urlencoded"),
                RequestMethod.POST,
                tokenRequest
        );

        try {
            var response = httpClient.send(oAuth2Request, HttpResponse.BodyHandlers.ofString());

            if (null != response && null != response.body()) {
                Map<String, Object> map
                        = objectMapper.readValue(response.body(), new TypeReference<>() {
                });

                tokenBearer = (String) map.get("access_token");
            } else {
                throw new TemplatingServiceException("Failed to get access token for template service");
            }
        } catch (IOException e) {
            throw new TemplatingServiceException("Error while trying to get auth token for template service", e);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new TemplatingServiceException("Error while trying to get auth token for template service", e);
        }
    }

    /**
     * HTTP send request method
     * <p>
     * This method exists so that the application can resend the request with a new access token in case of 401
     *
     * @param request
     * @param bodyHandler
     * @param <T>
     * @return
     */
    private <T> HttpResponse<T> makeRequest(HttpRequest request, HttpResponse.BodyHandler<T> bodyHandler) throws WebServiceException {
        setupHttpClient();

        try {

            var response = httpClient.send(request, bodyHandler);

            if (response.statusCode() != 200) {
                getAccessToken();
                response = httpClient.send(request, bodyHandler);

                if (response.statusCode() != 200) {
                    throw new WebServiceException("Failed to access File Service");
                }
            }

            return response;

        } catch (IOException e) {
            throw new TemplatingServiceException("Failed to send or receive request through HTTP client");
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new TemplatingServiceException("The thread was interrupted. Failed to access Templating Service");
        }
    }

    // HTTP Header builders

    /**
     * Override of {@link #buildHttpRequestHeader(String, Map, RequestMethod, String)}
     *
     * @param url
     * @param headerMap
     * @param requestMethod
     * @return
     */
    private HttpRequest buildHttpRequestHeader(String url, Map<String, String> headerMap, RequestMethod
            requestMethod) {
        return buildHttpRequestHeader(url, headerMap, requestMethod, null);
    }

    /**
     * This method exists to make it easier to build the headers of the HTTP request
     * It also validates the requestMethod
     *
     * @param url
     * @param headerMap
     * @param requestMethod
     * @param bodyContent
     * @return
     */
    private HttpRequest buildHttpRequestHeader(String url, Map<String, String> headerMap, RequestMethod
            requestMethod, String bodyContent) {
        var httpRequest = HttpRequest.newBuilder(URI.create(url));

        if (null != headerMap && !headerMap.isEmpty()) {
            headerMap.keySet().forEach(key -> httpRequest.setHeader(key, headerMap.get(key)));
        }

        switch (requestMethod) {
            case GET:
                return httpRequest.GET().build();
            case POST:
                return httpRequest.POST(HttpRequest.BodyPublishers.ofString(bodyContent)).build();
            default:
                throw new TemplatingServiceException("HTTP Request method not recognized: " + requestMethod);
        }
    }

}
